maintenance_plan/models/maintenance.py

```diff
@@ -0,0 +1,222 @@
+
+
+class MaintenanceKind(models.Model):
+
+    _name = 'maintenance.kind'
+    _description = 'Maintenance Kind'
+
+    name = fields.Char('Name', required=True, translate=True)
+    active = fields.Boolean('Active Kind', required=True, default=True)
+
+    _sql_constraints = [
+        ('name_uniq', 'unique (name)',
+         "Maintenance kind name already exists !")]
+
+
+class MaintenancePlan(models.Model):
+
+    _name = 'maintenance.plan'
+    _description = 'Maintenance Plan'
+
+    equipment_id = fields.Many2one(string='Equipment',
+                                   comodel_name='maintenance.equipment',
+                                   ondelete='cascade')
+    maintenance_kind_id = fields.Many2one(string='Maintenance kind',
+                                          comodel_name='maintenance.kind',
+                                          ondelete='restrict')
+
+    period = fields.Integer(string='Period',
+                            help='Days between each maintenance')
+    duration = fields.Float(string='Duration',
+                            help='Maintenance duration in hours')
+
+    next_maintenance_date = fields.Date('Next maintenance date',
+                                        compute='_compute_next_maintenance')
+
+    @api.depends('period', 'maintenance_kind_id',
+                 'equipment_id.maintenance_ids.request_date',
+                 'equipment_id.maintenance_ids.close_date',
+                 'equipment_id.maintenance_ids.maintenance_kind_id')
+    def _compute_next_maintenance(self):
+
+        date_now = fields.Date.context_today(self)
+        today_date = fields.Date.from_string(date_now)
+
+        for plan in self.filtered(lambda x: x.period > 0):
+
+            period_timedelta = timedelta(days=plan.period)
+
+            next_maintenance_todo = self.env['maintenance.request'].search([
+                ('equipment_id', '=', plan.equipment_id.id),
+                ('maintenance_type', '=', 'preventive'),
+                ('maintenance_kind_id', '=', plan.maintenance_kind_id.id),
+                ('stage_id.done', '!=', True),
+                ('close_date', '=', False)], order="request_date asc", limit=1)
+            last_maintenance_done = self.env['maintenance.request'].search([
+                ('equipment_id', '=', plan.equipment_id.id),
+                ('maintenance_type', '=', 'preventive'),
+                ('maintenance_kind_id', '=', plan.maintenance_kind_id.id),
+                ('stage_id.done', '=', True),
+                ('close_date', '!=', False)], order="close_date desc", limit=1)
+            if next_maintenance_todo and last_maintenance_done:
+                next_date = next_maintenance_todo.request_date
+                date_gap = fields.Date.from_string(
+                    next_maintenance_todo.request_date) - \
+                    fields.Date.from_string(last_maintenance_done.close_date)
+                # If the gap between the last_maintenance_done and the
```

Comment:
@elicoidal Thanks for your review.

This code is an adaptation from odoo to keep the behaviour as close as possible to standard module : https://github.com/odoo/odoo/blob/10.0/addons/maintenance/models/maintenance.py#L165